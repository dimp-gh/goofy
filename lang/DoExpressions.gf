module DoExpressions where

# do-expression consists of a bunch of statements (separated with commas)
# and a resulting expression.
fun foo () = do {
  do println "simple do example:",
  do print "  ",
  fun bar x = + x, # you can define local functions inside do-expressions
  println (bar 5 10)
}

# nested do-expressions are fine (expect they look cumbersome)
fun nestedDos () = do {
  do println "nested do-expressions are fine:",
  do println "  outside1",
  do (do {
    do println "  inside1",
    println "  inside2"
  }),
  println "  outside2"    
}

fun main () = do {
  do foo (),
  nestedDos ()
}