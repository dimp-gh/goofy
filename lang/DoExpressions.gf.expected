simple do example:
  15
nested do-expressions are fine:
  outside1
  inside1
  inside2
  outside2
